### PR TITLE
RT-1.11 - BGP remove private AS  - OTG test

### PR DESCRIPTION
**Pull request by cprabha**
Hi,

I have added OTG test for RT-1.11 - BGP remove private AS.
Resolves : https://github.com/openconfig/featureprofiles/issues/1813

Thanks,
Prabha

### CODE REVIEW

**Comment by octpetre** [comment]
@cprabha could you please add this test to the PR header: _Resolves #1813_ 
This will ensure tracking of the OTG tests dev found here https://github.com/orgs/openconfig/projects/3/views/1

**Comment by cprabha** [comment]
> @cprabha could you please add this test to the PR header: _Resolves #1813_ This will ensure tracking of the OTG tests dev found here https://github.com/orgs/openconfig/projects/3/views/1

This is resolved.

**Review by sachendras — APPROVED** [review]

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go
Comment (jasdeep-hundal) [inline feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go:1]: File should be renamed to bgp_remove_private_as_test.go

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+	nbrs := []*bgpNeighbor{nbr1v4, nbr2v4}
+
+	d := &oc.Root{}
+	ni1 := d.GetOrCreateNetworkInstance(deviations.DefaultNetworkInstance(dut))
+	ni_proto := ni1.GetOrCreateProtocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_BGP, "BGP")
+	bgp := ni_proto.GetOrCreateBgp()
+	global := bgp.GetOrCreateGlobal()
+	global.RouterId = ygot.String(dutDst.IPv4)
+	global.As = ygot.Uint32(localAs)
+	global.GetOrCreateAfiSafi(oc.BgpTypes_AFI_SAFI_TYPE_IPV4_UNICAST).Enabled = ygot.Bool(true)
```
Comment (jasdeep-hundal) [inline feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go:133]: ATE version also does the same for IPv6: https://github.com/openconfig/featureprofiles/blob/e3dc053e7c0abe36b6899660c377b7f0bbe09632/feature/experimental/bgp/ate_tests/bgp_remove_private_as/bgp_remove_private_as_test.go#L131

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+
+	// Note: we have to define the peer group even if we aren't setting any policy because it's
+	// invalid OC for the neighbor to be part of a peer group that doesn't exist.
+	pg1 := bgp.GetOrCreatePeerGroup(peerGrpName1)
+	pg1.PeerAs = ygot.Uint32(ateAS1)
+	pg1.PeerGroupName = ygot.String(peerGrpName1)
+
+	pg2 := bgp.GetOrCreatePeerGroup(peerGrpName2)
+	pg2.PeerAs = ygot.Uint32(ateAS2)
+	pg2.PeerGroupName = ygot.String(peerGrpName2)
```
Comment (jasdeep-hundal) [inline feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go:143]: The ATE test includes some more deviation based configuration here: https://github.com/openconfig/featureprofiles/blob/e3dc053e7c0abe36b6899660c377b7f0bbe09632/feature/experimental/bgp/ate_tests/bgp_remove_private_as/bgp_remove_private_as_test.go#L143-L163

Is that no longer needed?

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+	statePath := gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut)).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_BGP, "BGP").Bgp()
+	for _, nbr := range nbrIP {
+		var status *ygnmi.Value[oc.E_Bgp_Neighbor_SessionState]
+		nbrPath := statePath.Neighbor(nbr)
+		t.Logf("Waiting for BGP neighbor to establish...")
+		status, ok := gnmi.Watch(t, dut, nbrPath.SessionState().State(), time.Minute, func(val *ygnmi.Value[oc.E_Bgp_Neighbor_SessionState]) bool {
+			state, ok := val.Val()
+			return ok && state == oc.Bgp_Neighbor_SessionState_ESTABLISHED
+		}).Await(t)
+		state, _ := status.Val()
```
Comment (jasdeep-hundal) [inline feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go:172]: This line should be after the `if !ok` check.

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+	iDut2Eth := iDut2Dev.Ethernets().Add().SetName(ateDst.Name + ".Eth").SetMac(ateDst.MAC)
+	iDut2Eth.Connection().SetChoice(gosnappi.EthernetConnectionChoice.PORT_NAME).SetPortName(port2.Name())
+	iDut2Ipv4 := iDut2Eth.Ipv4Addresses().Add().SetName(ateDst.Name + ".IPv4")
+	iDut2Ipv4.SetAddress(ateDst.IPv4).SetGateway(dutDst.IPv4).SetPrefix(uint32(ateDst.IPv4Len))
+	iDut2Ipv6 := iDut2Eth.Ipv6Addresses().Add().SetName(ateDst.Name + ".IPv6")
+	iDut2Ipv6.SetAddress(ateDst.IPv6).SetGateway(dutDst.IPv6).SetPrefix(uint32(ateDst.IPv6Len))
+
+	iDut1Bgp := iDut1Dev.Bgp().SetRouterId(iDut1Ipv4.Address())
+	iDut1Bgp4Peer := iDut1Bgp.Ipv4Interfaces().Add().SetIpv4Name(iDut1Ipv4.Name()).Peers().Add().SetName(ateSrc.Name + ".BGP4.peer")
+	iDut1Bgp4Peer.SetPeerAddress(iDut1Ipv4.Gateway()).SetAsNumber(ateAS1).SetAsType(gosnappi.BgpV4PeerAsType.EBGP)
```
Comment (jasdeep-hundal) [inline feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go:223]: Would prefer setting the peer address as `dutSrc.IPv4` and similarly for the other peer. (Keeps the intent of peering with the specific DUT interface clearer.)

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+	bgpNeti1Bgp4PeerRoutes.SetNextHopIpv4Address(iDut1Ipv4.Address()).
+		SetNextHopAddressType(gosnappi.BgpV4RouteRangeNextHopAddressType.IPV4).
+		SetNextHopMode(gosnappi.BgpV4RouteRangeNextHopMode.MANUAL)
+	bgpNeti1Bgp4PeerRoutes.Addresses().Add().
+		SetAddress(advertisedRoutesv4Net).
+		SetPrefix(uint32(advertisedRoutesv4Prefix)).
+		SetCount(routeCount)
+
+	if asSEQMode {
+		bgpNeti1Bgp4PeerRoutes.AsPath().SetAsSetMode(gosnappi.BgpAsPathAsSetModeEnum("include_as_seq"))
```
Comment (jasdeep-hundal) [inline feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go:243]: According to https://github.com/open-traffic-generator/fp-testbed-juniper/issues/3#issuecomment-1682099875, we should be able to set AS path segments to routes, can we add that here now?

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+	})
+
+	t.Run("Configure DEFAULT network instance", func(t *testing.T) {
+		dutConfNIPath := gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut))
+		gnmi.Replace(t, dut, dutConfNIPath.Type().Config(), oc.NetworkInstanceTypes_NETWORK_INSTANCE_TYPE_DEFAULT_INSTANCE)
+	})
+
+	dutConfPath := gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut)).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_BGP, "BGP")
+	t.Run("Configure BGP Neighbors", func(t *testing.T) {
+		gnmi.Delete(t, dut, dutConfPath.Config())
```
Comment (jasdeep-hundal) [inline feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go:362]: Need to configure route policy here like in the ATE test: https://github.com/openconfig/featureprofiles/blob/e3dc053e7c0abe36b6899660c377b7f0bbe09632/feature/experimental/bgp/ate_tests/bgp_remove_private_as/bgp_remove_private_as_test.go#L339

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+		gotPrefixCount := len(bgpPrefixes)
+		if gotPrefixCount != routeCount {
+			t.Errorf("Received prefixes on otg are not as expected got prefixes %v, want prefixes %v", gotPrefixCount, routeCount)
+		} else {
+			t.Logf("Received prefixes on otg are matched, got prefixes %v, want prefixes %v", gotPrefixCount, routeCount)
+		}
+	}
+
+	/*
+		// Below code will be un commented once issue 1659 is resolved for otg.
```
Comment (jasdeep-hundal) [inline feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go:317]: Can we uncomment this now?

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/metadata.textproto
Comment (jasdeep-hundal) [inline feature/experimental/bgp/otg_tests/bgp_remove_private_as/metadata.textproto:1]: This is different than the metadata used for the ATE test. I'm assuming this is more up-to-date, but requesting confirmation.

**Review by jasdeep-hundal — CHANGES_REQUESTED** [review]

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go:1]: Updated.

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go (hunk 1 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go]: updated, thank you

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go (hunk 2 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go]: updated and in line with ATE Test.

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go (hunk 4 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go]: updated for both the peers.

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go (hunk 5 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go]: updated and remove ticket details, todo details.

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go (hunk 6 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go]: updated, inline with ATE test.

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go (hunk 7 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go]: updated.

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/metadata.textproto
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_remove_private_as/metadata.textproto:1]: I have taken from other bgp tests,  I am not sure of the latest deviations for other vendor but taken care for Juniper. Please let me know if any pointers for latest deviations, will refer it and update.

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go (hunk 3 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_remove_private_as/bgp_remove_private_as_otg_test.go]: updated.

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/metadata.textproto
Comment (jasdeep-hundal) [reply feature/experimental/bgp/otg_tests/bgp_remove_private_as/metadata.textproto:1]: Is there any reason it shouldn't be identical to https://github.com/openconfig/featureprofiles/blob/main/feature/experimental/bgp/ate_tests/bgp_remove_private_as/metadata.textproto?

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/README.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+BGP remove private AS
+
+## Procedure
+
+*   Establish BGP sessions as follows between ATE and DUT .  
+*   ATE port 1 emulates two eBGP neighbors peering the DUT. 
+*   eBGP neighbor # 1 is injecting routes with AS_PATH modified to have private AS numbers 
+*   Validate that private AS numbers are stripped before advertisement to the eBGP neighbor on ATE 
+    port 2.
```
Comment (jasdeep-hundal) [inline feature/experimental/bgp/otg_tests/bgp_remove_private_as/README.md:13]: There's extra whitespace at the end of these lines, and the space before the period in the first line should be removed.

Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_remove_private_as/README.md]: Updated.

File: feature/experimental/bgp/otg_tests/bgp_remove_private_as/metadata.textproto
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_remove_private_as/metadata.textproto:1]: Made in line with ate metadata.

**Review by jasdeep-hundal — APPROVED** [review]

**Pull request merged**